In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
import tensorflow as tf
import random
%matplotlib inline

pd.set_option('display.max_rows', 200)
pd.set_option('display.max_columns', 200)

2023-10-31 10:40:12.944916: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [11]:

#Read dataset
train_a = pd.read_parquet('A/train_targets.parquet')
train_b = pd.read_parquet('B/train_targets.parquet')
train_c = pd.read_parquet('C/train_targets.parquet')
X_train_estimated_a = pd.read_parquet('A/X_train_estimated.parquet')
X_train_estimated_b = pd.read_parquet('B/X_train_estimated.parquet')
X_train_estimated_c = pd.read_parquet('C/X_train_estimated.parquet')
X_train_observed_a = pd.read_parquet('A/X_train_observed.parquet')
X_train_observed_b = pd.read_parquet('B/X_train_observed.parquet')
X_train_observed_c = pd.read_parquet('C/X_train_observed.parquet')

#add location to each sample
train_a["location"] = "A"
train_b["location"] = "B"
train_c["location"] = "C"
X_train_estimated_a["location"] = "A"
X_train_estimated_b["location"] = "B"
X_train_estimated_c["location"] = "C"
X_train_observed_a["location"] = "A"
X_train_observed_b["location"] = "B"
X_train_observed_c["location"] = "C"

#remove extra minute 00 sample
X_train_observed_a = X_train_observed_a.iloc[:-1,:]
X_train_observed_b = X_train_observed_b.iloc[:-1,:]
X_train_observed_c = X_train_observed_c.iloc[:-1,:]

#add date_calc column same as date_forecast column to observed data
X_train_observed_a.insert(0, "date_calc", X_train_observed_a["date_forecast"])
X_train_observed_b.insert(0, "date_calc", X_train_observed_b["date_forecast"])
X_train_observed_c.insert(0, "date_calc", X_train_observed_c["date_forecast"])

#concat all the samples and remove date_calc column
X_train_raw = pd.concat([X_train_observed_a,
                     X_train_observed_b,
                     X_train_observed_c,
                     X_train_estimated_a,
                     X_train_estimated_b,
                     X_train_estimated_c])

#feature indicating time between date_calc and date_forecast
X_train_raw["calc_time"] =(X_train_raw["date_forecast"] - X_train_raw["date_calc"]).astype('timedelta64[s]')

#fill nans
X_train_raw["snow_density:kgm3"] = X_train_raw["snow_density:kgm3"].apply(
    lambda a : np.isnan(a)
    ).map({True: 0, False: 1})
X_train_raw["ceiling_height_agl:m"] = X_train_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)
X_train_raw["cloud_base_agl:m"] = X_train_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)

#create seperate dataframes for measurments at minute 00, 15, 30 and 45
X_train00 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 0)].reset_index().iloc[:,1:]
X_train15 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 15)].reset_index().iloc[:,1:]
X_train30 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 30)].reset_index().iloc[:,1:]
X_train45 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 45)].reset_index().iloc[:,1:]

#remove redundant data
X_train15 = X_train15.iloc[:,2:-2]
X_train30 = X_train30.iloc[:,2:-2]
X_train45 = X_train45.iloc[:,2:-2]

#join observations into single sample
X_train = X_train00.join(X_train15, lsuffix="_00", rsuffix="_15").join(X_train30.join(X_train45, lsuffix="_30", rsuffix="_45"))

#rename column for merging with targets
X_train = X_train.rename(columns={"date_forecast" : "time"})

#concat target values and drop NaN values
targets = pd.concat([train_a,
                     train_b,
                     train_c]).dropna()

#merge weatherfeatures with corresponding target pv measurement
dataset = pd.merge(X_train, targets, how="right", on=["time", "location"])

#shuffle dataset
#dataset = dataset.sample(frac=1, random_state=43).reset_index().iloc[:,1:]

#sort by time
dataset = dataset.sort_values(by="time")

#split into features and targets
datasetX = dataset.iloc[:, :-1]
datasetY = dataset.iloc[:, -1:]

#add day_of_year and hour feature columns
datasetX["day"] = datasetX["time"].dt.day_of_year
datasetX["hour"] = datasetX["time"].dt.hour

#get indexes of samples in the months of the test dataset
indexMayJuneJuly = datasetX["time"].apply(lambda time : time.month in [5, 6, 7])

#OHE encoding for catagorical feature "location"
datasetX["location_A"] = datasetX["location"].apply(lambda a : a == "A").map({True: 1, False: 0})
datasetX["location_B"] = datasetX["location"].apply(lambda a : a == "B").map({True: 1, False: 0})
datasetX["location_C"] = datasetX["location"].apply(lambda a : a == "C").map({True: 1, False: 0})

datasetX = datasetX.drop("location", axis=1)

#move datecalc column
#date_calc_column = datasetX.pop("date_calc")
#datasetX.insert(183, "is_not_calculated", date_calc_column) #!constant 183 can be source of bugs

#fix nans for some reason???
#datasetX["is_not_calculated"] = datasetX["is_not_calculated"].fillna(method="ffill")

#map location labels to numbers
#datasetX["location"] = datasetX["location"].map({"A": 0, "B": 1, "C": 2})

#drop time and date_calc columns
datasetX = datasetX.iloc[:,2:]

#calculate mean and std for normalizing data, values should also be used for normalizing test data
dataMean = datasetX.mean()
dataStd = datasetX.std()

#normalize data
datasetX.iloc[:,:-4] = ((datasetX.iloc[:,:-4]-dataMean[:-4])/dataStd[:-4]).fillna(value=0)


evalsetSize = 10000
evalsetIndex = 80000
#partition into training and evalset
trainsetX = pd.concat([datasetX.iloc[:evalsetIndex,:], datasetX.iloc[evalsetIndex+evalsetSize:,:]])
trainsetY = pd.concat([datasetY.iloc[:evalsetIndex,:], datasetY.iloc[evalsetIndex+evalsetSize:,:]])
trainsetIndexMayJuneJuly = indexMayJuneJuly[:85000]
evalsetX = datasetX.iloc[evalsetIndex:evalsetIndex+evalsetSize,:]
evalsetY = datasetY.iloc[evalsetIndex:evalsetIndex+evalsetSize,:]
evalsetIndexMayJuneJuly = indexMayJuneJuly[85000:]

display(datasetX)

KeyboardInterrupt: 

In [13]:
numModels = 5

models = []

for i in range(numModels):
    models.append(tf.keras.models.Sequential([
        #tf.keras.layers.GaussianNoise(stddev=0.05, seed=42),
        tf.keras.layers.Dense(100, activation="relu", kernel_initializer=tf.keras.initializers.GlorotNormal),
        tf.keras.layers.Dense(80, activation="relu", kernel_initializer=tf.keras.initializers.GlorotNormal),
        tf.keras.layers.Dense(60, activation="relu", kernel_initializer=tf.keras.initializers.GlorotNormal),
        tf.keras.layers.Dense(1, activation="relu", kernel_initializer=tf.keras.initializers.GlorotNormal),
    ]))

In [14]:
i = 0
while(i < numModels):
    models[i].compile(
        optimizer=tf.keras.optimizers.experimental.Adadelta(learning_rate=1, 
                                                            #weight_decay=0.0001
                                                            ),
        loss="mean_absolute_error"
    )

    history = models[i].fit(x = trainsetX,
                        y = trainsetY,
                        batch_size = 1000,
                        epochs=100,
                        verbose=0
                    )
    
    #retry if model is nonsensical
    loss = models[i].evaluate(evalsetX, evalsetY)
    if(loss > 200):
        continue
    
    print(i, " complete")
    i+=1

313/313 [==============================] - 1s 2ms/step - loss: 69.7636
0  complete
313/313 [==============================] - 1s 2ms/step - loss: 69.6413
1  complete
313/313 [==============================] - 1s 4ms/step - loss: 74.3601
2  complete
313/313 [==============================] - 1s 2ms/step - loss: 70.5042
3  complete
313/313 [==============================] - 1s 2ms/step - loss: 206.0912


KeyboardInterrupt: 

In [5]:
#!ReWrite
def evaluate_models(models, X, Y):
    preds = X.iloc[:,1:2]

    for i in range(len(models)):
        preds[str(i)] = models[i].predict(X)
        
    
    preds = preds.iloc[:,1:]
    
    
    preds["final"] = preds.mean(axis=1)
    preds["losses"] = (preds["final"] - Y["pv_measurement"]).apply(lambda a : np.abs(a))
    return preds["losses"].mean()


In [12]:

print("entire:")
results = evaluate_models(models ,evalsetX, evalsetY)
print(results)



entire:
313/313 [==============================] - 1s 2ms/step
68.82488570192321


131.8, 30.6, 125.4, 111.9, 41.0, 152.5
124.9, 24.5, 118.7, 107.9, 42.0, 145.3
????  ????

In [ ]:
print("may june july:")
resultsMayJuneJuly = evaluate_models(models ,evalsetX[evalsetIndexMayJuneJuly], evalsetY[evalsetIndexMayJuneJuly])
print(resultsMayJuneJuly)



In [ ]:
print("location A")
resultA = evaluate_models(models ,evalsetX[evalsetX["location_A"].apply(lambda a : a == 1)], 
                         evalsetY[evalsetX["location_A"].apply(lambda a : a == 1)])
print(resultA)

print("location B")
resultB = evaluate_models(models ,evalsetX[evalsetX["location_B"].apply(lambda a : a == 1)], 
                         evalsetY[evalsetX["location_B"].apply(lambda a : a == 1)])
print(resultB)

print("location C")
resultC = evaluate_models(models ,evalsetX[evalsetX["location_C"].apply(lambda a : a == 1)], 
                         evalsetY[evalsetX["location_C"].apply(lambda a : a == 1)])
print(resultC)



In [10]:
print("Small samplesize 1")
resultSmall1 = evaluate_models(models, evalsetX.iloc[:1300,:], evalsetY.iloc[:1300,:])
print(resultSmall1)

print("Small samplesize 2")
resultSmall2 = evaluate_models(models, evalsetX.iloc[1300:2600,:], evalsetY.iloc[1300:2600,:])
print(resultSmall2)

print("Small samplesize 3")
resultSmall3 = evaluate_models(models, evalsetX.iloc[2600:3900,:], evalsetY.iloc[2600:3900,:])
print(resultSmall3)

Small samplesize 1
41/41 [==============================] - 0s 3ms/step
248.31683480096405
Small samplesize 2
41/41 [==============================] - 0s 3ms/step
88.25757709596759
Small samplesize 3
41/41 [==============================] - 0s 3ms/step
36.94612070169197


67    68

249.9 248.3
90.5  88.2
33.0  36.9

71.5  69.6
152.4  148.7
152.5 148.9
33.2 32.0
17.6 16.9

71.5  72.4  72.4
152.4 153.3 152.7
152.5 155.0 155.0
33.2  33.1  33.1
17.6  17.5  17.5

With 50% training data: tanh
76.5  72.0
161.7 158.6
164.0 164.9
34.6  35.2
18.7  18.7
74.2  74.2
82.1  84.6

With 50% training data: relu


In [86]:
X_test_estimated_a = pd.read_parquet('A/X_test_estimated.parquet')
X_test_estimated_b = pd.read_parquet('B/X_test_estimated.parquet')
X_test_estimated_c = pd.read_parquet('C/X_test_estimated.parquet')

X_test_estimated_a["location"] = "A"
X_test_estimated_b["location"] = "B"
X_test_estimated_c["location"] = "C"

X_test_raw = pd.concat([
                     X_test_estimated_a,
                     X_test_estimated_b,
                     X_test_estimated_c])

#feature indicating time between date_calc and date_forecast
X_test_raw["calc_time"] =(X_test_raw["date_forecast"] - X_test_raw["date_calc"]).astype('timedelta64[s]')

#fill nans
X_test_raw["snow_density:kgm3"] = X_test_raw["snow_density:kgm3"].apply(
    lambda a : np.isnan(a)
    ).map({True: 0, False: 1})
X_test_raw["ceiling_height_agl:m"] = X_test_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)
X_test_raw["cloud_base_agl:m"] = X_test_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)

#create seperate dataframes for measurments at minute 00, 15, 30 and 45
X_test00 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 0)].reset_index().iloc[:,1:]
X_test15 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 15)].reset_index().iloc[:,1:]
X_test30 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 30)].reset_index().iloc[:,1:]
X_test45 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 45)].reset_index().iloc[:,1:]

#remove redundant data
X_test15 = X_test15.iloc[:,2:-2]
X_test30 = X_test30.iloc[:,2:-2]
X_test45 = X_test45.iloc[:,2:-2]

#join observations into single sample
X_test_estimated = X_test00.join(X_test15, lsuffix="_00", rsuffix="_15").join(X_test30.join(X_test45, lsuffix="_30", rsuffix="_45"))


X_test_estimated = X_test_estimated.rename(columns={"date_forecast" : "time"})

parse_dates = ['time']
X_test_targets = pd.read_csv("test.csv", parse_dates=parse_dates)

X_test = pd.merge(X_test_estimated, X_test_targets, on=["time", "location"], how="right").iloc[:,:-2]


X_test["day"] = X_test["time"].dt.day_of_year
X_test["hour"] = X_test["time"].dt.hour

X_test["location_A"] = X_test["location"].apply(lambda a : a == "A").map({True: 1, False: 0})
X_test["location_B"] = X_test["location"].apply(lambda a : a == "B").map({True: 1, False: 0})
X_test["location_C"] = X_test["location"].apply(lambda a : a == "C").map({True: 1, False: 0})

X_test = X_test.drop("location", axis=1)

X_test = X_test.iloc[:,2:]


X_test.iloc[:,:-4] = ((X_test.iloc[:,:-4]-dataMean[:-4])/dataStd[:-4]).fillna(value=0)

display(X_test)



/var/folders/bm/37km6rb530l8y0vv0znc0zb00000gn/T/ipykernel_12419/1064356002.py:63: PerformanceWarning: Adding/subtracting object-dtype array to TimedeltaArray not vectorized.
  X_test.iloc[:,:-4] = ((X_test.iloc[:,:-4]-dataMean[:-4])/dataStd[:-4]).fillna(value=0)


absolute_humidity_2m:gm3_00  air_density_2m:kgm3_00  \
0                       -0.595904                0.834929   
1                       -0.632742                0.862243   
2                       -0.669581                0.780297   
3                       -0.706420                0.725662   
4                       -0.780097                0.725662   
...                           ...                     ...   
2155                     0.840800               -1.623536   
2156                     0.914477               -1.541587   
2157                     1.024993               -1.459637   
2158                     1.098670               -1.350374   
2159                     1.098670               -1.323056   

      ceiling_height_agl:m_00  clear_sky_energy_1h:J_00  clear_sky_rad:W_00  \
0                   -0.382669                 -0.627834           -0.626244   
1                   -1.083697                 -0.627834           -0.626244   
2                   -0.173977                 -0.627834           -0.626244   
3                    0.128249                 -0.627834           -0.626244   
4                    0.134297                 -0.554985           -0.438066   
...                       ...                       ...                 ...   
2155                 0.577483                  0.122098           -0.113350   
2156                 0.539586                 -0.299286           -0.451195   
2157                 0.197264                 -0.540086           -0.604801   
2158                 0.014887                 -0.626155           -0.626244   
2159                 0.017892                 -0.627834           -0.626244   

      cloud_base_agl:m_00  dew_or_rime:idx_00  dew_point_2m:K_00  \
0               -0.382669           -0.028554          -0.517627   
1               -1.083697           -0.028554          -0.532260   
2               -0.173977           -0.028554          -0.590784   
3                0.128249           -0.028554          -0.649314   
4                0.134297           -0.028554          -0.722471   
...                   ...                 ...                ...   
2155             0.577483           -0.028554           0.930882   
2156             0.539586           -0.028554           0.974774   
2157             0.197264           -0.028554           1.018671   
2158             0.014887           -0.028554           1.077196   
2159             0.017892           -0.028554           1.062563   

      diffuse_rad:W_00  diffuse_rad_1h:J_00  direct_rad:W_00  \
0            -0.651235            -0.658524        -0.444504   
1            -0.651235            -0.658524        -0.444504   
2            -0.651235            -0.658524        -0.444504   
3            -0.651235            -0.658524        -0.444504   
4            -0.133490            -0.396493        -0.347113   
...                ...                  ...              ...   
2155          0.029749             0.264405        -0.301073   
2156         -0.326408            -0.149611        -0.392267   
2157         -0.568792            -0.452422        -0.444504   
2158         -0.651235            -0.616985        -0.444504   
2159         -0.651235            -0.658524        -0.444504   

      direct_rad_1h:J_00  effective_cloud_cover:p_00  elevation:m_00  \
0              -0.449899                    0.390542       -0.685740   
1              -0.449899                   -0.071723       -0.685740   
2              -0.449899                    0.781470       -0.685740   
3              -0.449899                    0.227893       -0.685740   
4              -0.400659                   -0.240079       -0.685740   
...                  ...                         ...             ...   
2155           -0.307607                    0.644503        1.599325   
2156           -0.350643                    0.496121        1.599325   
2157           -0.423319                    0.050977        1.599325   
2158           -0.

In [87]:
#!ReWrite
def models_predict(models, X):
    preds = X.iloc[:,1:2]

    for i in range(len(models)):
        preds[str(i)] = models[i].predict(X)

    preds = preds.iloc[:,1:]
    preds["final"] = preds.mean(axis=1)
    return preds["final"]

In [89]:
test_preds = pd.read_csv("sample_submission.csv")
test_preds["prediction"] = models_predict(models, X_test)
display(test_preds)
test_preds.to_csv("preds/Ensamble3.csv", index=False)

68/68 [==============================] - 0s 2ms/step


id  prediction
0        0    0.000000
1        1    0.000000
2        2    0.000000
3        3   53.867607
4        4  388.250275
...    ...         ...
2155  2155   53.368217
2156  2156   22.887728
2157  2157    0.000000
2158  2158    0.000000
2159  2159    0.000000

[2160 rows x 2 columns]